### PR TITLE
OAK-8902 | Listing lucene blobs using oak-run

### PR DESCRIPTION
**Pull request by nit0906**

### CODE REVIEW

File: oak-run/src/main/java/org/apache/jackrabbit/oak/run/DataStoreCheckCommand.java [hunk 1]
```diff
@@ -242,16 +248,23 @@ static int checkDataStore(String... args) {
...
                     NodeTraverser traverser = new NodeTraverser(nodeStore, dsType);
                     closer.register(traverser);
-                    traverser.traverse();
+
+                    List<String> rootPathList = options.valuesOf(verboseRootPath);
+                    traverser.traverse((String[]) rootPathList.toArray(new String[rootPathList.size()]));
+
                     FileUtils.copyFile(traverser.references, register.createFile(refOp, dumpPath));
                 } else {
                     retrieveBlobReferences(blobStore, marker,
```
Comment (thomasmueller) [inline oak-run/src/main/java/org/apache/jackrabbit/oak/run/DataStoreCheckCommand.java:270]: So, as far as I understand, for SegmentStore, it just traverses the nodes, but for all other stores (MongoDB / DocumentStore), it gets all the references of all the nodes... Right? That would be a lot of references... I think it should traverse just the nodes in verboseRootPath in both cases, if verboseRootPath is set.

File: oak-run/src/main/java/org/apache/jackrabbit/oak/run/DataStoreCheckCommand.java [hunk 2]
```diff
@@ -242,16 +248,23 @@ static int checkDataStore(String... args) {
...
 
             if (options.has(refOp) || options.has(consistencyOp)) {
-                if (options.has(verbose) &&
+
+                // Find blob ids by traversal for verbose mode + Segment Store or if verboseRootPath option
+                // is present (find blob references under a specific root path.)
+                if ((options.has(verbose) &&
                     (nodeStore instanceof SegmentNodeStore ||
-                        nodeStore instanceof org.apache.jackrabbit.oak.segment.SegmentNodeStore)) {
+                        nodeStore instanceof org.apache.jackrabbit.oak.segment.SegmentNodeStore)) ||
```
Comment (thomasmueller) [inline oak-run/src/main/java/org/apache/jackrabbit/oak/run/DataStoreCheckCommand.java]: Ah I see, old and new segment store...

**Review by thomasmueller — CHANGES_REQUESTED** [review]

File: oak-run/src/main/java/org/apache/jackrabbit/oak/run/DataStoreCheckCommand.java (hunk 1 above)
Comment (thomasmueller) [reply oak-run/src/main/java/org/apache/jackrabbit/oak/run/DataStoreCheckCommand.java:270]: Ah sorry, I see now! Looks good, I just didn't understand it.

**Review by thomasmueller — APPROVED** [review]

File: oak-run/src/test/java/org/apache/jackrabbit/oak/run/DataStoreCommandTest.java [hunk 3]
```diff
@@ -322,6 +325,71 @@ public void testConsistencyVerbose() throws Exception {
         testConsistency(dump, data, true);
     }
 
+    @Test
+    public void testConsistencyVerboseRootPathNoMissing() throws Exception {
```
Comment (fabriziofortino) [inline oak-run/src/test/java/org/apache/jackrabbit/oak/run/DataStoreCommandTest.java:329]: Codacy Issue found: [JUnit tests should include assert() or fail()](https://app.codacy.com/gh/oak-indexing/jackrabbit-oak/pullRequest?prid=5026544)

File: oak-run/src/test/java/org/apache/jackrabbit/oak/run/DataStoreCommandTest.java [hunk 4]
```diff
@@ -322,6 +325,71 @@ public void testConsistencyVerbose() throws Exception {
...
+        for (String id : data.idToPath.keySet()) {
+            if (data.idToPath.get(id).equals("/c1") || data.idToPath.get(id).equals("/c2")) {
+                data.addedSubset.add(id);
+            }
+        }
+        testConsistency(dump, data, true, true);
+    }
+
+    @Test
+    public void testConsistencyVerboseRootPath() throws Exception {
```
Comment (fabriziofortino) [inline oak-run/src/test/java/org/apache/jackrabbit/oak/run/DataStoreCommandTest.java:361]: Codacy Issue found: [JUnit tests should include assert() or fail()](https://app.codacy.com/gh/oak-indexing/jackrabbit-oak/pullRequest?prid=5026544)

File: oak-run/src/test/java/org/apache/jackrabbit/oak/run/DataStoreCommandTest.java [hunk 5]
```diff
@@ -322,6 +325,71 @@ public void testConsistencyVerbose() throws Exception {
...
+        for (String id : data.idToPath.keySet()) {
+            if (data.idToPath.get(id).equals("/c1") || data.idToPath.get(id).equals("/c2")) {
+                data.addedSubset.add(id);
+            }
+        }
+        testConsistency(dump, data, true, true);
+    }
+
+    @Test
+    public void testConsistencyVerboseRootPathWithDeleted() throws Exception {
```
Comment (fabriziofortino) [inline oak-run/src/test/java/org/apache/jackrabbit/oak/run/DataStoreCommandTest.java:345]: Codacy Issue found: [JUnit tests should include assert() or fail()](https://app.codacy.com/gh/oak-indexing/jackrabbit-oak/pullRequest?prid=5026544)

File: oak-run/src/test/java/org/apache/jackrabbit/oak/run/DataStoreCommandTest.java [hunk 6]
```diff
@@ -353,6 +421,26 @@ public void gc() throws Exception {
...
+
+        additionalParams += " --verboseRootPath /a";
+        try {
+            testGc(dump, data, 0, false);
+        } catch (OptionException e) {
+            assertTrue(e.getMessage().equals("Option(s) [verboseRootPath] are unavailable " +
+                    "given other options on the command line"));
+            return;
+        }
+        assertFalse(true);
```
Comment (fabriziofortino) [inline oak-run/src/test/java/org/apache/jackrabbit/oak/run/DataStoreCommandTest.java:442]: Codacy Issue found: [assertTrue(true) or similar statements are unnecessary](https://app.codacy.com/gh/oak-indexing/jackrabbit-oak/pullRequest?prid=5026544)

File: oak-run/src/main/java/org/apache/jackrabbit/oak/run/DataStoreOptions.java [hunk 7]
```diff
@@ -76,6 +79,11 @@ public DataStoreOptions(OptionParser parser) {
         verbose =
             parser.accepts("verbose", "Option to get all the paths and implementation specific blob ids");
 
+        // Option NOT available for garbage collection operation - we throw an exception if both --collect-garbage and
+        // --verboseRootPath are provided in the command.
+        verboseRootPath = parser.accepts("verboseRootPath",
+                "Root path to output backend formatted ids/paths").availableUnless(collectGarbage).availableIf(verbose).withRequiredArg().withValuesSeparatedBy(DELIM).ofType(String.class);;
```
Comment (fabriziofortino) [inline oak-run/src/main/java/org/apache/jackrabbit/oak/run/DataStoreOptions.java:85]: Codacy Issue found: [An empty statement (semicolon) not part of a loop](https://app.codacy.com/gh/oak-indexing/jackrabbit-oak/pullRequest?prid=5026544)

**Pull request closed without merging**